autoload/go/util.vim

```diff
@@ -99,7 +99,13 @@ endfunction
 function! go#util#System(str, ...) abort
   let l:shell = &shell
   if !go#util#IsWin() && executable('/bin/sh')
-    let &shell = '/bin/sh'
+    " Having the [t]csh shells set as the login shell results in temporary
+    " file access errors when calling ksh or dash system shells. /bin/sh is
+    " often aliased to these shells on modern Linux and BSD systems including
+    " VoidLinux, Debian and FreeBSD.
+    if match(l:shell, 'csh$') == -1
+      let &shell = '/bin/sh'
+    endif
```

Comment:
The shellredir is the right way to go. Expanding on the deep dive option mentioned above:

The 'shellredir' option controls what is captured by system(). From Vim help page: "The default is ">".  For Unix, if the 'shell' option is "csh", "tcsh" or "zsh" during initializations, the default becomes ">&".  If the 'shell' option is "sh", "ksh" or "bash" the default becomes ">%s 2>&1".  This means that stderr is also included.

The initialization of this option is done after reading the ".vimrc" and the other initializations, so that when the 'shell' option is set there, the 'shellredir' option changes automatically unless it was explicitly set before."

So from #988  the shell was changed without updating shellredir to match which causes the error for the csh/tcsh/zsh users.

I have tested this successfully and will update the pull request.